### DISCUSSION TITLE
How to correctly migrate from V5 Mainline mode to V6 Mainline versioning without changes in behaviour?

### DISCUSSION THREAD

**Discussion by hotchkj** [GitTools/GitVersion#4143] [Q&A]
I have hundreds of repos configured for 5.12 with a config that is as follows:

``` yml
mode: Mainline
assembly-versioning-scheme: MajorMinor
branches: {}
ignore:
  sha: []
merge-message-formats: {}
```

Lets' say I have a `master` branch with a commit tagged with `v10.0.0` and eight more untagged commits after it, each of which had `+semver: none` in the commit message and correspondingly the build versioning skipped these as still producing `v10.0.0` with no change in tag. GitVersion is being run at `HEAD` of `master`.

GitVersion `5.12.0` gives me `10.0.0` with a pre-release tag of `8` which gets skipped when forming `FullSemVer`:

<details>
  <summary>JSON</summary>

``` json
{
  "Major": 10,
  "Minor": 0,
  "Patch": 0,
  "PreReleaseTag": "8",
  "PreReleaseTagWithDash": "",
  "PreReleaseLabel": "",
  "PreReleaseLabelWithDash": "",
  "PreReleaseNumber": null,
  "WeightedPreReleaseNumber": 60000,
  "BuildMetaData": null,
  "BuildMetaDataPadded": "",
  "FullBuildMetaData": "Branch.master.Sha.8947dafe06011c812599eb2b273830d916d367cf",
  "MajorMinorPatch": "10.0.0",
  "SemVer": "10.0.0",
  "LegacySemVer": "10.0.0",
  "LegacySemVerPadded": "10.0.0",
  "AssemblySemVer": "10.0.0.0",
  "AssemblySemFileVer": "10.0.0.0",
  "FullSemVer": "10.0.0",
  "InformationalVersion": "10.0.0+Branch.master.Sha.8947dafe06011c812599eb2b273830d916d367cf",
  "BranchName": "master",
  "EscapedBranchName": "master",
  "Sha": "8947dafe06011c812599eb2b273830d916d367cf",
  "ShortSha": "8947daf",
  "NuGetVersionV2": "10.0.0",
  "NuGetVersion": "10.0.0",
  "NuGetPreReleaseTagV2": "",
  "NuGetPreReleaseTag": "",
  "VersionSourceSha": "bf8dcfafbf235762f6e1d9396392dfac9ee690b6",
  "CommitsSinceVersionSource": 8,
  "CommitsSinceVersionSourcePadded": "0008",
  "UncommittedChanges": 327,
  "CommitDate": "2024-07-18"
}
```
</details>

which is correct - the package is unchanged.

If I try the suggested migration for 6.0.0, using trunk-based, I get completely different behaviour, so I'm querying - is there actually a 1:1 expected configuration for the above? Is the migration configuration advice incomplete or perhaps yielding assumptions?

``` yml
workflow: TrunkBased/preview1
assembly-versioning-scheme: MajorMinor
```

GitVersion `6.0.0` gives me `10.0.8`:

<details>
  <summary>JSON</summary>

``` json
{
  "AssemblySemFileVer": "10.0.8.0",
  "AssemblySemVer": "10.0.0.0",
  "BranchName": "master",
  "BuildMetaData": null,
  "CommitDate": "2024-07-18",
  "CommitsSinceVersionSource": 1,
  "EscapedBranchName": "master",
  "FullBuildMetaData": "Branch.master.Sha.8947dafe06011c812599eb2b273830d916d367cf",
  "FullSemVer": "10.0.8",
  "InformationalVersion": "10.0.8+Branch.master.Sha.8947dafe06011c812599eb2b273830d916d367cf",
  "Major": 10,
  "MajorMinorPatch": "10.0.8",
  "Minor": 0,
  "Patch": 8,
  "PreReleaseLabel": "",
  "PreReleaseLabelWithDash": "",
  "PreReleaseNumber": null,
  "PreReleaseTag": "",
  "PreReleaseTagWithDash": "",
  "SemVer": "10.0.8",
  "Sha": "8947dafe06011c812599eb2b273830d916d367cf",
  "ShortSha": "8947daf",
  "UncommittedChanges": 328,
  "VersionSourceSha": "f327e3cc4c5b92b77dd186198d2bb4262c1473fd",
  "WeightedPreReleaseNumber": 60000
}
```
</details>

which is incorrect; the package hasn't changed so the version should not have done so either.

What is the migration path to yield original behaviour? My expectation is that the versioning should be driven by the last tag, particularly since the interposed commits are all marked with the `none` semver message.

I cannot find any up-to-date documentation on either `Mainline` as a strategy (or indeed what any of the strategies actually do), and the documented `TrunkBased` strategy does not appear to exist.

Assistance/advice would be appreciated. I am suspecting I need to replace `strategies` but since I clearly don't grasp what these are actually doing, I would like some help in understanding the functionality.

What I am attempting to achieve:
- I want a trunk-based approach where the last-seen tag on one of the trunk branches is used to determine the next version
- If `master` was last tagged with `1.2.0`, then absent any other commit messaging, the next version should be `1.2.1` i.e. defaulting to a patch; if the next commit contains `+semver: feature` then it should yield `1.3.0`; our build system will then apply a tag accordingly
- We use `+semver: none` to indicate that the version should not change, and correspondingly no tag should be added
- Hypothetically if a tag were to be skipped (a build breakage of some kind) then it is preferred if the versioning scheme recognised the gap i.e. the following should still work
  - Commit A (v1.2.0)
  - Commit B +semver: none (no tag)
  - Commit C (no tag but should have been v1.2.1)
  - Commit D (v1.2.2)
  - Thus rebuilding C should still yield the right tag
- If a `support` branch was last tagged with `1.3.2` then the next version should be `1.3.3`
- Any other branches/PRs should be a pre-release of some kind

**Comment by arturcic** [comment]
@HHobeck can you help here?

**Reply by HHobeck** [reply]
Actually the new `Mainline` version strategy is a new implementation and has nothing to do with the old one. You will find the following sentence in the breaking change documentation:

> The Mainline mode and the related implementation has been removed completely. The new Mainline version strategy should be used instead.

It is not clear mentioned in this document but the new implementation has not the goal to be backward compatible and might have a different behavior compared to the old version. What I know is, the `+semver: none` has no effect when using it as a bump message. I have thought about it and came to the conclusion we need to implement this behavior maybe by using `=semver: none` bump message.

Edit:
To answer your question: No there is no guidance to migrate to the new Mainline strategy.

**Reply by HHobeck** [reply]
@hotchkj: I have created a feature ticket for that. Please feel free and implement it

- https://github.com/GitTools/GitVersion/issues/4144

**Reply by hotchkj** [reply]
Thanks @HHobeck - is there any way with current strategies to re-create something close to what I described, in the interim? Instead of using `strategies: Mainline`, borrowing from GitFlow or GitHubFlow strategies?

I'm trying to gauge whether there's a path forward with `6.0.0` or not at this time - we're at a juncture internally where as part of infrastructure work. either we stay on `5.12` for a year or two further and just accept that NET 6 tooling goes out of support in November, or upgrade*, but that's contingent on having a viable approach we can roll out on `6.0.0`. Changing hundreds of `GitVersion.yml` files is not something we can do often.

I can consult with our legal on whether we can devote any time to this enhancement, but the above infrastructure project will be taking priority.

_\* I suppose there is also the option of recompiling 5.12 for NET 8 but I'd rather not do that unless the GitVersion team consider doing so!_

**Reply by HHobeck** [reply]
> Thanks @HHobeck - is there any way with current strategies to re-create something close to what I described, in the interim?

If you don't want the behavior that the version is incrementing (in your example to patch) on every commit you need to just specify the increment to `none`:

```yaml
workflow: TrunkBased/preview1
branches:
  main:
    increment: none
```

> Instead of using strategies: Mainline, borrowing from GitFlow or GitHubFlow strategies?

The same answer applies when using the `GitFlow/v1` or other workflows. Like I said `+semver: none` bump messages has no effects for all strategies.

> I'm trying to gauge whether there's a path forward with 6.0.0 or not at this time - we're at a juncture internally where as part of infrastructure work. either we stay on 5.12 for a year or two further and just accept that NET 6 tooling goes out of support in November, or upgrade*, but that's contingent on having a viable approach we can roll out on 6.0.0. Changing hundreds of GitVersion.yml files is not something we can do often.

Good point. That is a question you need to answer yourself. The new workflow pattern gives you the possibility to switch very easily from one to another workflow. Of course only when following the default workflow behavior. Like you see our `TrunkBased/preview1` workflow is experimental and needs maybe improvements (from the configuration point of view). Please feel free and create a change request if you think the workflow is not correct configured.

> I can consult with our legal on whether we can devote any time to this enhancement, but the above infrastructure project will be taking priority.

Yes please do. Commercial companies and others need to not only take from open source projects they need to give also something back. As a maintainer I would appreciate it.

**Reply by hotchkj** [reply]
I think we do need more context, however.

>The same answer applies when using the GitFlow/v1 or other workflows. Like I said +semver: none bump messages has no effects for all strategies.

How/why did Mainline mode work the way it did in 5.12, if this does not apply? Why was it abandoned? We need some assurance that there is a pathway forward supported by the GitVersion team to restoring that functionality.

Is it feasible to get pointers to where this was removed in 6, to try and see what needs restoring? I'm trying to grasp if there are further features and changes needed beyond the one listed, to get back to Mainline mode behaviour. After all, this would need to be a maintained workflow in future, that needs to be well understood.

For example, I might want to add integration tests that pass under 5.12 first.

**Reply by HHobeck** [reply]
> How/why did Mainline mode work the way it did in 5.12, if this does not apply? Why was it abandoned?

Again:

**The Mainline version strategy implementation in version 6.0.0 (which will be used in the TrunkBased/preview1 workflow) is a new implementation and has nothing to do with the old Mainline mode implementation in 5.12. The new implementation has not the goal to be backward compatible.**

The previous algorithm was error prone and not maintainable. Thus the initial plan of the maintainers of GitVersion was to delete the algorithm completely without any alternative. Also no one has used it (at least no one we are aware of). Because I personally want to save the workflow I decided to re-implement it from scratch and invest docent of hours in my spare time.

> We need some assurance that there is a pathway forward supported by the GitVersion team to restoring that functionality.

There is no assurance. But I invite you to contribute.

> Is it feasible to get pointers to where this was removed in 6, to try and see what needs restoring?

If you like you can analyze the previous code and try to rescue it. But restoring of an old algorithm is not the right way in my opinion. Better understand the new algorithm and implement the feature based on acceptance criterias.

> I'm trying to grasp if there are further features and changes needed beyond the one listed, to get back to Mainline mode behavior. After all, this would need to be a maintained workflow in future, that needs to be well understood.

The only features I know of are the following:

- #4057
- #4144

But to be honest. I have no idea what features where included in the previous algorithm.

**At the end I would like to mention that all the time the maintainer and contributer are investing here in this project is unpaid and voluntary. Please keep that in mind before expressing any requirements.**

**Reply by hotchkj** [reply]
> Because I personally want to save the workflow I decided to re-implement it from scratch and invest docent of hours in my spare time.

That is very much appreciated. I think it would have been less confusing had it received a different name, as my query would have opened with '_how do we get something like Mainline back?_'. I would be interested, perhaps in a separate thread once other concerns are set aside, to grasp the intent and see how closely it fits.

> There is no assurance. But I invite you to contribute.

That's not how open source maintenance works. You as maintainers have to determine the agreed scope of the project - the architectural impact, the feature sets desired to maintain, what is championed as the capabilities of GitVersion. External contributors cannot contribute to a project that is hostile to the concept they want to contribute, so it needs to be understood whether a contribution fits in. This is exactly why forking ends up occurring.

If there's no assurance that a contribution won't simply get deleted next week because a maintainer disagrees with it, how can anyone contribute in good faith?

> Also no one has used it (at least no one we are aware of).

How would you know? I find this assertion curious - does GitVersion do surveys? Data gathering? Is there something we should have participated in? Did you scan github.com for `GitVersion.yml` files? What decision process was used here? Does GitVersion have an intended feature set? How would a repeat of this be avoided, say, for GitVersion 7? These are the sorts of questions I have to be able to answer before getting investment made.

> But to be honest. I have no idea what features where included in the previous algorithm.

I don't know the history of this - but consider - if this algorithm was implemented by someone previously involved with GitVersion and now deleted because they've stopped said involvement - this implicitly sends the message that contributions can be removed later if nobody champions them. That would mean you're not asking for contributions; you're asking for additional maintainers, and that is a _very important_ distinction to make in such discussions, since I need to convey this to various leadership.

> At the end I would like to mention that all the time the maintainer and contributer are investing here in this project is unpaid and voluntary. Please keep that in mind before expressing any requirements.

This is well understood and does not need repeating. Volunteer time on all sorts of open source projects (including my own) is always laudable and appreciated. It does not change that if you want external contributions, as maintainers it has to be clear whether those contributions fit your intended scope. Let me give an example from another recent issue from @arturcic 

> Second option will not be accepted by the team

This indicates the team has discussed contributions and agreed that some will not be accepted.

I ask again - are changes like the ones I describe in line with your intended direction for GitVersion? What would keep that workflow alive in future? Presumably _someone_ liked Mainline mode and then it got removed. What prevents future investment being similarly removed?

This is a great tool, one that it's clear a lot of people benefit from, and the team should be proud of what has been accomplished. It perhaps might benefit from greater clarity on intent.

> Please keep that in mind before expressing any requirements.

I'd invite you to consider the tone of this discussion. I haven't expressed requirements; I have asked questions and am attempting to gauge whether the decisions made in GitVersion 6 are compatible with potential future contributions.

In a word, help me (figure out if I can) help you 😄

**Reply by HHobeck** [reply]
> That's not how open source maintenance works. You as maintainers have to determine the agreed scope of the project - the architectural impact, the feature sets desired to maintain, what is championed as the capabilities of GitVersion. External contributors cannot contribute to a project that is hostile to the concept they want to contribute, so it needs to be understood whether a contribution fits in. This is exactly why forking ends up occurring.

I thought you are asking about an assurance that someone else is implementing it for you. I agree the maintainers need to give the direction or path and help with architectural design decisions.

Thus if you asking me the maintainers should:

- review incoming bugs and features 
- decide whether it is compatible with existing concepts (triage process)
- answering question and ideas in the discussion forum
- helping out with architectural design decisions

If the feature or bug has been accepted then it can be solved (by contributor or maintainer) and merged to the production code.

In your concrete example I have created a feature ticket and documented the acceptance criteria’s. You can now review it and check if it fits your needs. Because it is not labeled with the triage flag and created from a maintainer you can be sure it is accepted and aligns with the overall design of GitVersion. What do you want more?

> I don't know the history of this - but consider - if this algorithm was implemented by someone previously involved with GitVersion and now deleted because they've stopped said involvement - this implicitly sends the message that contributions can be removed later if nobody champions them. That would mean you're not asking for contributions; you're asking for additional maintainers, and that is a very important distinction to make in such discussions, since I need to convey this to various leadership.

Ideally every feature needs to have an acceptance test which ensures the functionality. If you as a contributor want to be sure that your feature remains I would recommend you to write integration tests and document it on the web page.

The Mainline mode algorithm lives from the idea in the Mainline version strategy of course. Because of the fact all acceptance tests which were present are still existing and green. This tiny part with pump message `+semver: none` was changed and aligned with the other algorithmic behavior. This was a design decision of maintainers you need to accept. If you want this behavior back please go ahead or wait until someone is implementing it. You can think of it like: it is work in progress. That is the reason why the Mainline version strategy is marked with "preview usage only".

> I ask again - are changes like the ones I describe in line with your intended direction for GitVersion? What would keep that workflow alive in future? Presumably someone liked Mainline mode and then it got removed. What prevents future investment being similarly removed?

Actually your requirement is additive. If you take a look into the feature #4144 you will see a different regular expression to detect this kind of messages (`=semver: none` instead of `+semver: none`). Because it will be in the configuration you can change it and use it like in the old days. It fits very well in the current design and yes I see no reason not implementing it this way.

> I'd invite you to consider the tone of this discussion. I haven't expressed requirements; I have asked questions and am attempting to gauge whether the decisions made in GitVersion 6 are compatible with potential future contributions.
> 
> In a word, help me (figure out if I can) help you 😄

Yes that is very kind thank you. I think with the refactoring initiative from 5.0 to 6.0 the code is much more maintainable. Together with the massive number of integration tests GitVersion has -- with its implemented functionalities -- a bright future. :)

**Reply by arturcic** [reply]
@hotchkj, because you mentioned me on my response on the other discussion, I will jump in.

> That's not how open source maintenance works. You as maintainers have to determine the agreed scope of the project - the architectural impact, the feature sets desired to maintain, what is championed as the capabilities of GitVersion. External contributors cannot contribute to a project that is hostile to the concept they want to contribute, so it needs to be understood whether a contribution fits in. This is exactly why forking ends up occurring.

First of all I don't think forking is a bad idea in open source. I actually think it's a good idea because it can serve as a playground for new ideas that can later at some point get integrated into the original repository (or maybe not, but still serves as an ideas sharing mechanism).

I can agree that if you as a maintainer want to have good contributions from externals you have to communicate somehow what's the vision medium to long term of the project. And that's what would `assure` the potential contributor his effort will not be for nothing. That's the perfect way to do open source.

But, that's not the case in our (and many more) project(s). There are many reasons for that - it's a small project, a small team, and not that many potential contributors willing to invest any time (probably because of the reasons above, but most of them just do not want to spend time at all, and most of them are pretty vocal requesting changes).

Now, sometimes we do get some pretty good contributions that fit our vision (that I think has to be vocalized better) and most of them get merged into the main repository. But then there is the catch - these contributions have to be maintained by us, as the original authors are not willing to continue on maintaining their contribution. So long terms it adds up the burden on the maintainers.

A good example would be the `GitVersion.MsBuild`, you mentioned my comment from one of the discussions related to that package. The biggest problem is maintaining compatibility with every new sdk and old frameworks. That is really hard for us to spend that amount of time keeping it up to date and making sure it works. We had to invest a lot of time on integrations tests and e2e tests done in Docker. But that really slows down any other development for us (.net Framework is a good way to slow us down) Now you mentioned  
> does GitVersion do surveys? Data gathering?

We do not do extensive research as big teams/big projects do, mainly because we are a really small team, that spends much of time on maintaining the code. I would really want to do that extensive research to drive the long terms goals of the project, but at least from the moment I joined the team the main driver was how the maintainers use the tool and the most active contributors.

Regarding the small team - there are currently 2 main active maintainers - me and @HHobeck (which joined recently, like 1 year or a bit more). We have 2 more maintainers - from the initial maintainers that whenever they have time, help us with code reviews, or whenever we need to consult them. And take in consideration how much time 2+ people can allocate on maintaining the code and other activities (like vocalizing the medium/long terms goals and helping the community that sometimes is really rude).

Back to the data gathering/ metrics. We are following the usage of variant distros we publish: GitVersion.Tool and derivates (Chocolatey, winget, homebrew, docker images, GitHub actions, Azure Pipelines) and GitVersion.MsBuild and the effort we put into improving/maintaining each of the deliverables. You know what's interesting - we spend like 65% on the GitVersion.MsBuild (fixes, questions) and 30% on the Core, and the rest for the GitVersion.Tool (and derivates). Now in terms of downloads/usages GitVersion.Tool is the leader - 90% of downloads, with docker/winget/homebrew (which can be included in the GitVersion.Tool as they derivate from it) - 5% and GitVersion.MsBuild - the rest 5%. So I guess it speaks for itself where we should invest our time in.

As for the configurations changes that @HHobeck introduced in v6 (major - means breaks stuff). 
> How/why did Mainline mode work the way it did in 5.12, if this does not apply? Why was it abandoned.

I think @HHobeck explained, but I can add my cents - the configuration module in place in v6 had to change in a way that is more clear how it works and how to extend/customize. Maybe we need more documentations (like use case scenarios and what should be changed to achieve the wanted use case) , that will be really nice. But the main effort was put on bringing all the "flows" under the same structure and predictable behavior (with many more tests). The way it worked in 5.12 was a bit messy and with different approaches to similar problems. 

> Why was it abandoned.

And as @HHobeck said, it's not abandoned, it's just we need time/people willing to get involved on medium/long term.

Now if we talk about the GitVersion 7.0 and later, the goals are:
* to refactor the cli to make it more in line with the accepted expectations from a modern cli 
* have a new implementation of the "git" layer - replacing the native Libgit2Sharp with a managed solution
* adapt the codebase to the new versions of dotnet/C#
* prepare the codebase to support AoT
* On the configuration module - re-introduce the `gitversion init` that will be maintainable and in sync with the changes we do in the configuration module, and maybe additional flows.

Sorry I have to go the long way on explaining, but I will later use this post as a reference for other people asking the same questions.

Have a good day.

**Reply by hotchkj** [reply]
Thanks both for the detailed replies, very useful context! Perfectly understood on the lengthy replies, hopefully others will find them useful as well.

I will file the paperwork our end to discuss potential contributions. At this point I can't promise anything, but as we have some vested interest, my hope is that someone on our teams will be able to assist. However, it's got to go through legal etc. first which may take some time...

**Reply by HHobeck** [reply]
@hotchkj: May I ask if the paperwork was successful?

**Reply by hotchkj** [reply]
@HHobeck still in Legal land at present 😞 it can be a slow process

**Reply by koppor** [reply]
Since the update of https://github.com/gittools/actions, this becomes more important again. May I ping you @hotchkj regarding the legal paper work? I was with Mercedes-Benz AG (we won an award regarding our training) and Nokia Bell Labs, maybe, I can assist.

**Reply by hotchkj** [reply]
@koppor to be clear, the paperwork is on my side for contributions - if anyone wants to tackle restoring the mainline functionality to GitVersion, they needn't wait on me per se. I can also test changes vs our current process easily enough, it's the signoff from legal on what I can spend time on coding that prevents my contributing.

**Reply by koppor** [reply]
>  it's the signoff from legal on what I can spend time on coding that prevents my contributing.

I worked with legal in my former roles and therefore I offered support. I can also connect to some other open source experts if it needs to be with more formal roles (in addition to experience). This is what I meant to offer.

**Reply by koppor** [reply]
@hotchkj we send you some energy for the legal paperwork

**Comment by strickie** [comment]
@hotchkj / @koppor What did you end up doing?
We are figthing (a loosing) battle with GitVersion v6 - Every try we have attempted, is just a litlle off. Almosy every on of our hundreds repos have a little somthing each, that make a common strategy almost impossible... :-(
Any help is much appreciated.

**Reply by koppor** [reply]
Still on my TODO list - we seem to have messed up 10 years ago when introducing GitButler - config at https://github.com/JabRef/jabref/blob/main/GitVersion.yml

I have some Claude subscription and I plan to use it to work on this.

**Reply by HHobeck** [reply]
> We are figthing (a loosing) battle with GitVersion v6 - Every try we have attempted, is just a litlle off. Almosy every on of our hundreds repos have a little somthing each, that make a common strategy almost impossible... :-(
> Any help is much appreciated.

It sounds like you’re dealing with a lot of edge cases across repositories. Could you share a couple of concrete examples of what’s not working as expected? Are you looking for the feature #4144?
